File: backend/src/routes/cofounder.ts
import { Router } from "express";

import { db } from "../db/index.js";
import { cofounderMatches } from "../db/schema.js";
import { cofounderSearchRequest } from "../schemas.js";
import { cofounderMatchOut } from "../serialize.js";
import * as pipeline from "../services/pipeline.js";
import { getProfileOr404, touchProfile } from "../services/profiles.js";
import * as webClient from "../services/webClient.js";

const SYNTHESIS_SYSTEM_PROMPT = `You are a cofounder-matching assistant. Given a \
founder's own profile, the complementary profile they're looking for, and raw \
candidate signal from the web, produce a JSON array of up to 5 candidates: \
{name, headline, profile_url, match_rationale, skill_tags: []}. \
match_rationale must explain the *complementary* fit (why this candidate \
covers what the founder is missing), not just similarity. Respond with a \
JSON array only, no markdown fences.`;

type RawCofounderMatch = {
  name?: string;
  headline?: string | null;
  profile_url?: string | null;
  match_rationale?: string;
  skill_tags?: string[];
};

/** Plain web search for now -- the richer LinkedIn people-search dataset
 * tool (brightdataClient.runDatasetTool) is on hold until Bright Data
 * access is unblocked, see bright-data.md. */
function gatherCandidates(domain: string, desiredComplement: string): Promise<unknown> {
  const query = `${desiredComplement} cofounder ${domain}`;
  return pipeline.cachedFetch({
    toolName: "search_engine",
    query,
    fetchFn: () => webClient.searchEngine(query),
  });
}

export const cofounderRouter = Router();

cofounderRouter.post("/search", async (req, res) => {
  const payload = cofounderSearchRequest.parse(req.body);
  const profile = await getProfileOr404(payload.profile_id);

  // 1. gather
  const raw = await gatherCandidates(profile.domain || "startup", payload.desired_complement);

  // 2. extract
  const facts = await pipeline.extractFacts(
    JSON.stringify(raw).slice(0, 8000),
    `Extract candidate people (name, headline, profile url) matching: ${payload.desired_complement}`,
  );

  // 3. synthesize
  const synthesis = await pipeline.synthesize(
    SYNTHESIS_SYSTEM_PROMPT,
    `Founder profile: ${JSON.stringify(payload.founder_profile)}\n` +
      `Looking for: ${payload.desired_complement}\n\n` +
      `Candidate signal:\n${JSON.stringify(facts)}`,
  );
  const matchesData = pipeline.parseJson<RawCofounderMatch[]>(synthesis, []);

  await touchProfile(profile.id, { founderSkills: payload.founder_profile });

  const rows = matchesData.map((match) => ({
    profileId: profile.id,
    name: match.name ?? "",
    headline: match.headline ?? null,
    profileUrl: match.profile_url ?? null,
    matchRationale: match.match_rationale ?? "",
    skillTags: match.skill_tags ?? [],
  }));
  const inserted = rows.length ? await db.insert(cofounderMatches).values(rows).returning() : [];

  res.json({ matches: inserted.map(cofounderMatchOut) });
});

File: backend/app/routers/cofounder.py
import json

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from app.db import get_db
from app.models import CofounderMatch, StartupProfile
from app.schemas import CofounderSearchRequest, CofounderSearchResponse
from app.services import pipeline, web_client

router = APIRouter(prefix="/cofounder", tags=["cofounder"])

SYNTHESIS_SYSTEM_PROMPT = """You are a cofounder-matching assistant. Given a \
founder's own profile, the complementary profile they're looking for, and raw \
candidate signal from the web, produce a JSON array of up to 5 candidates: \
{name, headline, profile_url, match_rationale, skill_tags: []}. \
match_rationale must explain the *complementary* fit (why this candidate \
covers what the founder is missing), not just similarity. Respond with a \
JSON array only, no markdown fences."""


def _gather_candidates(db: Session, domain: str, desired_complement: str) -> list[dict]:
    """Plain web search for now -- the richer LinkedIn people-search dataset
    tool (brightdata_client.run_dataset_tool) is on hold until Bright Data
    access is unblocked, see bright-data.md."""
    query = f"{desired_complement} cofounder {domain}"
    return pipeline.cached_fetch(
        db,
        tool_name="search_engine",
        query=query,
        fetch_fn=lambda: web_client.search_engine(query),
    )


@router.post("/search", response_model=CofounderSearchResponse)
def search_cofounders(payload: CofounderSearchRequest, db: Session = Depends(get_db)):
    profile = db.get(StartupProfile, payload.profile_id)
    if not profile:
        raise HTTPException(status_code=404, detail="profile not found")

    # 1. gather
    raw = _gather_candidates(db, profile.domain or "startup", payload.desired_complement)

    # 2. extract
    facts = pipeline.extract_facts(
        raw_context=json.dumps(raw)[:8000],
        extraction_prompt=f"Extract candidate people (name, headline, profile url) matching: {payload.desired_complement}",
    )

    # 3. synthesize
    synthesis = pipeline.synthesize(
        system_prompt=SYNTHESIS_SYSTEM_PROMPT,
        user_prompt=(
            f"Founder profile: {json.dumps(payload.founder_profile)}\n"
            f"Looking for: {payload.desired_complement}\n\n"
            f"Candidate signal:\n{json.dumps(facts)}"
        ),
    )
    try:
        matches_data = json.loads(synthesis)
    except ValueError:
        matches_data = []

    profile.founder_skills = payload.founder_profile
    matches = []
    for m in matches_data:
        match = CofounderMatch(
            profile_id=profile.id,
            name=m.get("name", ""),
            headline=m.get("headline"),
            profile_url=m.get("profile_url"),
            match_rationale=m.get("match_rationale", ""),
            skill_tags=m.get("skill_tags", []),
        )
        db.add(match)
        matches.append(match)
    db.commit()
    for match in matches:
        db.refresh(match)
    return CofounderSearchResponse(matches=matches)
